File: stereo-camera/movie3d.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/calib3d/calib3d.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <stdio.h>
#include <iostream>


int main()
{
  // Check for left and right camera IDs
  std::string CamL_id{"data/stereoL.mp4"}, CamR_id{"data/stereoR.mp4"}; // These values can change depending on the system

  cv::VideoCapture camL(CamL_id), camR(CamR_id);
  
  cv::Mat Left_Stereo_Map1, Left_Stereo_Map2;
  cv::Mat Right_Stereo_Map1, Right_Stereo_Map2;

  cv::FileStorage cv_file = cv::FileStorage("data/params_cpp.xml", cv::FileStorage::READ);
  cv_file["Left_Stereo_Map_x"] >> Left_Stereo_Map1;
  cv_file["Left_Stereo_Map_y"] >> Left_Stereo_Map2;
  cv_file["Right_Stereo_Map_x"] >> Right_Stereo_Map1;
  cv_file["Right_Stereo_Map_y"] >> Right_Stereo_Map2;
  cv_file.release();

  // Check if left camera is attched
  if (!camL.isOpened())
  {
    std::cout << "Could not open camera with index : " << CamL_id << std::endl;
    return -1;
  }

  // Check if right camera is attached
  if (!camL.isOpened())
  {
    std::cout << "Could not open camera with index : " << CamL_id << std::endl;
    return -1;
  }

  cv::Mat frameL, frameR;
  
  for (size_t i{0}; i<100000; i++)
  {
    camL >> frameL;
    camR >> frameR;

    cv::Mat Left_nice, Right_nice;

    cv::remap(frameL,
              Left_nice,
              Left_Stereo_Map1,
              Left_Stereo_Map2,
              cv::INTER_LANCZOS4,
              cv::BORDER_CONSTANT,
              0);

    cv::remap(frameR,
              Right_nice,
              Right_Stereo_Map1,
              Right_Stereo_Map2,
              cv::INTER_LANCZOS4,
              cv::BORDER_CONSTANT,
              0);


    cv::Mat Left_nice_split[3], Right_nice_split[3];

    std::vector<cv::Mat> Anaglyph_channels;

    cv::split(Left_nice, Left_nice_split);
    cv::split(Right_nice, Right_nice_split);

    Anaglyph_channels.push_back(Left_nice_split[0]);
    Anaglyph_channels.push_back(Left_nice_split[1]);
    Anaglyph_channels.push_back(Right_nice_split[2]);

    cv::Mat Anaglyph_img;

    cv::merge(Anaglyph_channels, Anaglyph_img);

    cv::imshow("Anaglyph image", Anaglyph_img);
    cv::waitKey(1);
  }
  
  return 0;
}

File: stereo-camera/movie3d.py
import numpy as np 
import cv2


CamL_id = "data/stereoL.mp4"
CamR_id = "data/stereoR.mp4"

CamL= cv2.VideoCapture(CamL_id)
CamR= cv2.VideoCapture(CamR_id)

print("Reading parameters ......")
cv_file = cv2.FileStorage("data/params_py.xml", cv2.FILE_STORAGE_READ)

Left_Stereo_Map_x = cv_file.getNode("Left_Stereo_Map_x").mat()
Left_Stereo_Map_y = cv_file.getNode("Left_Stereo_Map_y").mat()
Right_Stereo_Map_x = cv_file.getNode("Right_Stereo_Map_x").mat()
Right_Stereo_Map_y = cv_file.getNode("Right_Stereo_Map_y").mat()
cv_file.release()


while True:
	retR, imgR= CamR.read()
	retL, imgL= CamL.read()
	
	if retL and retR:
		imgR_gray = cv2.cvtColor(imgR,cv2.COLOR_BGR2GRAY)
		imgL_gray = cv2.cvtColor(imgL,cv2.COLOR_BGR2GRAY)

		Left_nice= cv2.remap(imgL,Left_Stereo_Map_x,Left_Stereo_Map_y, cv2.INTER_LANCZOS4, cv2.BORDER_CONSTANT, 0)
		Right_nice= cv2.remap(imgR,Right_Stereo_Map_x,Right_Stereo_Map_y, cv2.INTER_LANCZOS4, cv2.BORDER_CONSTANT, 0)

		output = Right_nice.copy()
		output[:,:,0] = Right_nice[:,:,0]
		output[:,:,1] = Right_nice[:,:,1]
		output[:,:,2] = Left_nice[:,:,2]

		# output = Left_nice+Right_nice
		output = cv2.resize(output,(700,700))
		cv2.namedWindow("3D movie",cv2.WINDOW_NORMAL)
		cv2.resizeWindow("3D movie",700,700)
		cv2.imshow("3D movie",output)

		cv2.waitKey(1)
	
	else:
		break